accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -621,6 +614,488 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+  }
+
+  {
+    // final the same
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    } // Error, x is not assigned on all paths
+  }
+
+  { // late final the same
+    late x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // x has type int
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      if (b) {
+        x = 3;
+      } else {
+        x = 3;
+      }
+      x.isEven; // No error
+    } else {
+      x = "hello";
+      x.length; // No error
+    } // Error: type cannot be inferred for x
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+
+    if (b) {
+      if (b) {
+        x = 3;
+        return;
+      } else {
+        return;
+      } // x is inferred to int
+    } else {
+      x = 3;
+    } // x is inferred to int
+    x.isEven; // No error
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+
+    if (b) {
+      if (b) {
+        x = 3;
+        return;
+      } else {
+        x = null;
+        return;
+      } // x is inferred as int?
+    } else {
+      x = 3;
+    } // x is inferred as int?
+  }
+
+  {
+    var z; // z is an untyped variable
+    if (b) {
+      z = 3; // z is promoted to int
+     } else {
+      z = 3.0; // z is promoted to double
+      return;
+    } // Error, type cannot be inferred for z
+  }
+
+  {
+    var z; // z is an untyped variable
+    if (b) {
+      z = 3; // z is promoted to int
+      return;
+     } else {
+      z = 3.0; // z is promoted to double
+      return;
+    } // Error, type cannot be inferred for z
+  }
+
+  {
+    var u; // u is an untyped variable
+    if (b) {
+      u = 0 as num; // u is promoted to num
+      if (u is! int) u = 1;
+      // u has inferred type num, but is promoted to int
+    } else {
+      u = 3;
+      // u is inferred as int
+    } Error: type cannot be inferred for u
+  }
+}
+```
+
+It is an error to read an **untyped variable** when it has no **initialization
+inferred type**.
+
+It is an error to write to an **untyped variable** when it has no
+**initialization inferred type**, unless the write is an **initializing
+assignment**.
+
+
+```dart
+void test(bool b) {
+  {
+    var x; // x is an untyped variable
+    print(x); // Error, x is used before it is defined
+  }
+
+  {
+    late var x; // x is an untyped variable
+    var f = () => print(x); // Error, type cannot be inferred for x
+  }
+
+  {
+    var x; // x is an untyped variable
+    if (b) {
+      print(x); // Error: x has no initialization inferred type.
+      x = 3;
+    } else {
+      x = 4;
+    } // x is inferred as type int
+    print(x); // No error here, since x has been inferred at this point
+  }
+
+  {
+    var z;
+    do {
+      // z is potentially assigned.
+      if (b) {
+        z = 42;
+      } else {
+        z = 37;
+      }
+    } while (something);
+    print(z); // Error, type cannot be inferred for z
```

Comment:
The `do` loop is actually a bit more analyzable than a `while` loop because it is guaranteed to run at least once. So if the body considered without the back-edge gives rise to a typing, it should be sound to use that. The fact that it may repeat will still need to be taken into account, of course, such that we'll get an error if the loop body is writing to a `final` variable. But that might just be an extra run of the flow analysis on the loop body in a state where `z` is considered potentially assigned, checking for those "don't reassign" constraints.

Reply:
Personally I would rather not complicate flow analysis further by adding a special rule for `do` loops, especially if that means that we have to run flow analysis twice on the loop body.  It was an explicit design goal of flow analysis to make sure we didn't have to run it twice, since the CFE and analyzer resolution logic that it's fitting into is designed in such a way that doesn't permit running twice.  IMHO, `do` loops are rare enough in practice that the extra complexity we would take on by making this change wouldn't carry enough user benefit to justify it.